### ISSUE TITLE
Upgrade the spring-amqp version to work with spring-boot 1.5.3.RELEASE

### ISSUE THREAD

**Issue by marekdominiak** [spring-cloud/spring-cloud-contract#303]
My setup
- Spring boot: 1.5.3.RELEASE, with spring-boot-starter-amqp 
- Spring cloud dependencies: Dalston.RELEASE
- Spring cloud contract: 1.1.0.RELEASE

I got the issue with running the generated producer messaging tests (RabbitMQ, SpringAMQP). As far as I understand the @AutoConfigureMessageVerifier should replace the real connection factory with mockedRabbitMQ connection factory. While debugging the code I saw that the RabbitMockConnectionFactoryAutoConfiguration kicks in and creates the CachingConnetionFactory with an internal connection factory pointing to the mockedRabbitMQ connection factory. However while running the tests despite the fact that the Mocked version is used, somehow the RabbitTemplate convertAndSend method is trying to connect to the real RabbitMQ broker. 

This results in this NPE:
```
2017-05-19 10:09:25.055  INFO 18789 --- [           main] o.s.c.c.v.tests.ContractVerifierTest     : Started ContractVerifierTest in 8.959 seconds (JVM running for 9.511)
Hibernate: insert into customer (aggregate_status, version, cost_center_name, crm_number, customer_status, name, aggregate_id) values (?, ?, ?, ?, ?, ?, ?)
Hibernate: select customer0_.aggregate_id as aggregat1_1_0_, customer0_.aggregate_status as aggregat2_1_0_, customer0_.version as version3_1_0_, customer0_.cost_center_name as cost_cen4_1_0_, customer0_.crm_number as crm_numb5_1_0_, customer0_.customer_status as customer6_1_0_, customer0_.name as name7_1_0_ from customer customer0_ where customer0_.aggregate_id=?
2017-05-19 10:09:25.224  INFO 18789 --- [           main] o.s.a.r.c.CachingConnectionFactory       : Created new connection: SimpleConnection@36109a76 [delegate=null, localPort= 0]
2017-05-19 10:09:25.229  INFO 18789 --- [           main] o.s.a.r.c.CachingConnectionFactory       : Created new connection: SimpleConnection@497cf31a [delegate=null, localPort= 0]

java.lang.NullPointerException
    at org.springframework.amqp.rabbit.connection.SimpleConnection.createChannel(SimpleConnection.java:55)
    at org.springframework.amqp.rabbit.connection.CachingConnectionFactory$ChannelCachingConnectionProxy.createBareChannel(CachingConnectionFactory.java:1124)
    at org.springframework.amqp.rabbit.connection.CachingConnectionFactory$ChannelCachingConnectionProxy.access$200(CachingConnectionFactory.java:1113)
    at org.springframework.amqp.rabbit.connection.CachingConnectionFactory.doCreateBareChannel(CachingConnectionFactory.java:539)
    at org.springframework.amqp.rabbit.connection.CachingConnectionFactory.createBareChannel(CachingConnectionFactory.java:522)
    at org.springframework.amqp.rabbit.connection.CachingConnectionFactory.getCachedChannelProxy(CachingConnectionFactory.java:492)
    at org.springframework.amqp.rabbit.connection.CachingConnectionFactory.getChannel(CachingConnectionFactory.java:485)
    at org.springframework.amqp.rabbit.connection.CachingConnectionFactory.access$1400(CachingConnectionFactory.java:98)
    at org.springframework.amqp.rabbit.connection.CachingConnectionFactory$ChannelCachingConnectionProxy.createChannel(CachingConnectionFactory.java:1129)
    at org.springframework.amqp.rabbit.core.RabbitTemplate.doExecute(RabbitTemplate.java:1435)
    at org.springframework.amqp.rabbit.core.RabbitTemplate.execute(RabbitTemplate.java:1411)
    at org.springframework.amqp.rabbit.core.RabbitTemplate.execute(RabbitTemplate.java:1387)
    at org.springframework.amqp.rabbit.core.RabbitAdmin.initialize(RabbitAdmin.java:500)
    at org.springframework.amqp.rabbit.core.RabbitAdmin$11.onCreate(RabbitAdmin.java:419)
    at org.springframework.amqp.rabbit.connection.CompositeConnectionListener.onCreate(CompositeConnectionListener.java:33)
    at org.springframework.amqp.rabbit.connection.CachingConnectionFactory.createConnection(CachingConnectionFactory.java:571)
    at org.springframework.amqp.rabbit.core.RabbitTemplate.doExecute(RabbitTemplate.java:1430)
    at org.springframework.amqp.rabbit.core.RabbitTemplate.execute(RabbitTemplate.java:1411)
    at org.springframework.amqp.rabbit.core.RabbitTemplate.send(RabbitTemplate.java:712)
    at com.mgp.api.customer.CustomerAndOrgStructureExternalPusher.sendEvent(CustomerAndOrgStructureExternalPusher.java:138)
    at com.mgp.api.customer.CustomerAndOrgStructureExternalPusher.on(CustomerAndOrgStructureExternalPusher.java:45)
    at com.mgp.api.customer.CustomerAndOrgStructureExternalPusher$$FastClassBySpringCGLIB$$486bafc3.invoke(<generated>)
    at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204)
    at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:669)
    at com.mgp.api.customer.CustomerAndOrgStructureExternalPusher$$EnhancerBySpringCGLIB$$f39dbb8f.on(<generated>)
    at com.mgp.ContractVerifierBase.emitCustomerUpdatedEvent(ContractVerifierBase.java:44)
    at org.springframework.cloud.contract.verifier.tests.ContractVerifierTest.validate_shouldProduceValidCustomerUpdatedEvent(Co
```

Downgrading the spring-amqp version to the 1.6.6.RELEASE makes the issue go away. That suggests that is the root cause of the problem.

I have forked the example I was playing with and by bumping up the spring boot version you will get this exception.
The fork https://github.com/marekdominiak/testing-asynchonous-interactions
Go inside of the product-service and run `./gradlew check bootRepackage --stacktrace`

**Comment by marcingrzejszczak** [comment]
@marekdominiak I'm analyzing the sample and I don't really understand yet why it's working in the first place :) So the Spring Cloud Contract library integrates with Spring AMQP but via the `ContractVerifierMessaging` interface. Example of usage - https://github.com/spring-cloud/spring-cloud-contract/blob/master/tests/samples-messaging-amqp/src/test/groovy/com/example/AmqpMessagingApplicationSpec.groovy#L108-L111 . That way automatically stuff get stubbed OOB. I don't recall ever even intending on making `RabbitmqTemplate` stubbed. I'll try to debug and analyze why it worked at all :P

**Comment by marcingrzejszczak** [comment]
I've debugged this and I was right. The fact that it used to work is completely accidental. Is it a feature or a bug? Don't really know ;)

In older versions of Rabbit, the RabbitTemplate used the `mockConnectionFactory.newConnection(ExecutorService)` method to retrieve the connection. We managed to mock exactly that method. For compatibility's sake, I'll update the mock so that it works also with the new versions of Rabbit. I'll need to do some hacking to make it work but so be it ;)
